openedx/core/djangoapps/catalog/utils.py

```diff
@@ -114,45 +119,145 @@ def munge_catalog_program(catalog_program):
 
-        return data if data else {}
-    else:
-        return {}
 
+def get_course_runs(user, course_keys):
+    """
+    Get course run data from the course catalog service if not available in cache.
+
+    Arguments:
+        course_keys ([String]): A list of Course key strings identifying the run whose data we want.
+        user (User): The user to authenticate as when making requests to the catalog service.
 
-def get_run_marketing_url(course_key, user):
-    """Get a course run's marketing URL from the course catalog service.
+    Returns:
+        dict of catalog course runs against course keys, empty if no data could be retrieved.
+    """
+    catalog_course_runs_against_course_keys = CatalogCacheUtility.get_cached_catalog_course_runs(
+        course_keys
+    )
+    if len(catalog_course_runs_against_course_keys.keys()) != len(course_keys):
+        missing_course_keys = CatalogCacheUtility.get_course_keys_not_found_in_cache(
+            course_keys, catalog_course_runs_against_course_keys.keys()
+        )
+        catalog_course_runs_against_course_keys.update(
+            get_and_cache_course_runs(user, missing_course_keys)
+        )
+
+    return catalog_course_runs_against_course_keys
+
+
+def get_run_marketing_url(user, course_key):
+    """
+    Get a course run's marketing URL from the course catalog service.
 
     Arguments:
-        course_key (CourseKey): Course key object identifying the run whose marketing URL we want.
+          course_key (String): Course key string identifying the run whose marketing URL we want.
+
+    Returns:
+          string, the marketing URL, or None if no URL is available.
+    """
+    course_marketing_urls = get_run_marketing_urls(user, [course_key])
+    return course_marketing_urls.get(course_key)
+
+
+def get_run_marketing_urls(user, course_keys):
+    """
+    Get course run marketing URLs from the course catalog service against course keys.
+
+    Arguments:
+        course_keys ([String]): A list of Course key strings identifying the run whose data we want.
         user (User): The user to authenticate as when making requests to the catalog service.
 
     Returns:
-        string, the marketing URL, or None if no URL is available.
+        dict of run marketing URLs against course keys
     """
-    course_run = get_course_run(course_key, user)
-    return course_run.get('marketing_url')
+    course_marketing_urls = {}
+    course_catalog_dict = get_course_runs(user, course_keys)
+    if not course_catalog_dict:
+        return course_marketing_urls
+
+    for course_key in course_keys:
+        if course_key in course_catalog_dict:
+            course_marketing_urls[course_key] = course_catalog_dict[course_key].get('marketing_url')
+
+    return course_marketing_urls
+
+
+class CatalogCacheUtility(object):
+    """
+    Non-instantiatable Class that contains all utility methods for Catalog Cache.
+    """
+    __metaclass__ = abc.ABCMeta
```

Comment:
@rlucioni I don't find any use-case of instantiating this class; hence, made it abstract.